SCF/scf_zott.in

```diff
@@ -0,0 +1,191 @@
+scf_type = 'diis'
+
+# Set defaults
+maxiter = 20
+E_conv = 1.0E-8
+D_conv = 1.0E-4
+
+# Integral generation from Psi4's MintsHelper
+wfn = psi4.new_wavefunction(mol, psi4.get_global_option('BASIS'))
+t = time.time()
+mints = MintsHelper(wfn.basisset())
+S = np.asarray(mints.ao_overlap())
+
+# Get nbf and ndocc for closed shell molecules
+nbf = S.shape[0]
+ndocc = wfn.nalpha()
+
+print('\nNumber of occupied orbitals: %d' % ndocc)
+print('Number of basis functions: %d' % nbf)
+
+# Run a quick check to make sure everything will fit into memory
+I_Size = (nbf ** 4) * 8.e-9
+print("\nSize of the ERI tensor will be %4.2f GB." % I_Size)
+
+# Estimate memory usage
+memory_footprint = I_Size * 1.5
+if I_Size > numpy_memory:
+    clean()
+    raise Exception("Estimated memory utilization (%4.2f GB) exceeds numpy_memory \
+                    limit of %4.2f GB." % (memory_footprint, numpy_memory))
+
+# Compute required quantities for SCF
+V = np.asarray(mints.ao_potential())
+T = np.asarray(mints.ao_kinetic())
+
+if scf_type == 'df' or scf_type == 'diis': 
+    # This is a slight hack of Psi4, it was never ment to do this.
+    scf_e, scf_wfn = energy('SCF', return_wfn=True)
+    dfobj = DFTensor(scf_wfn, "DF_BASIS_SCF") 
+    Qpq = np.asarray(dfobj.Qso())  # Equation 28
+
+# Build the density fitted version
+if scf_type == 'diis':
+    Idf = np.einsum('Qpq,Qrs->pqrs', Qpq, Qpq)
+
+if scf_type == 'scf': # or scf_type == 'diis':
+    #Build the conventional version
+    # Remember: Stop building this and use the DF tensor, this is just for a demonstration
+    I = np.asarray(mints.ao_eri())
+    print("\nThe norm between the g and density-fitted g tensor is %10.4f" % np.linalg.norm(Idf - I))
+    print('\nTotal time taken for integrals: %.3f seconds.' % (time.time() - t))
+
+t = time.time()
+
+# Build H_core
+H = T + V
+
+# Orthogonalizer A = S^(-1/2)
+A = mints.ao_overlap()
+A.power(-0.5, 1.e-14)
+A = np.asarray(A)
+
+# Calculate initial core guess
+Hp = A.dot(H).dot(A)
+e, C2 = np.linalg.eigh(Hp)
+C = A.dot(C2)
+Cocc = C[:, :ndocc]
+D = np.einsum('pi,qi->pq', Cocc, Cocc)
+
+print('\nTotal time taken for setup: %.3f seconds' % (time.time() - t))
+
+print('\nStart SCF iterations:\n')
+t = time.time()
+E = 0.0
+Enuc = mol.nuclear_repulsion_energy()
+Eold = 0.0
+
+if scf_type == 'diis' or scf_type == 'df':
+    Fock_list = []
+    DIIS_error = []
+
+for SCF_ITER in range(1, maxiter + 1):
+
+    if scf_type == 'df':
+        # Build Density fitted elements
+        #Xp = np.einsum('qrs,rs->q', Qpq, D)
+        Xp = np.dot(Qpq.reshape(Qpq.shape[0], -1), D.reshape(-1))
+    
+        #Zpqr = np.einsum('pqs,rs->pqr', Qpq, D)
+        Z1 = np.dot(Qpq.reshape(Qpq.shape[0] * Qpq.shape[1], -1), Cocc).reshape(Qpq.shape[0], Qpq.shape[1], -1)     
+        J = np.dot(Qpq.reshape(Qpq.shape[0], -1).T, Xp).reshape(S.shape[0],-1)
+        K  = np.dot(np.transpose(Z1, (1, 0, 2)).reshape(Z1.shape[1], -1), np.transpose(Z1, (0, 2, 1)).reshape(-1, Z1.shape[1]))
```

Comment:
Id try not to write too much on a single line, especially nested functions like this. Makes things very hard to read. In addition, you should note that a transpose + reshape forces a copy. It may be worth trying to only have a single copy instead of two.